Streaming PromQL engine: optionally fall back to Prometheus' engine if query is not supported

charleskorn opened the pull request:
#### What this PR does

This PR adds optional support for falling back to Prometheus' engine if the streaming PromQL engine does not support the query expression.

The fallback behaviour can be controlled with a CLI flag, but is enabled by default.

If fallback is required, the reason is logged and recorded in the metric `cortex_streaming_promql_engine_unsupported_queries_total`.

#### Which issue(s) this PR fixes or relates to

(none)

#### Checklist

- [x] Tests updated.
- [n/a] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

jdbaldry commented on cmd/mimir/config-descriptor.json:
```diff
@@ -1887,6 +1887,17 @@
+          "kind": "field",
+          "name": "enable_promql_engine_fallback",
+          "required": false,
+          "desc": "If true, fall back to Prometheus' PromQL engine if the streaming engine is in use and it does not support a query.",
```
I tried to reword to make it more clear that it's a per-query fallback and not that a single unsupported query might make all subsequent queries use the Prometheus PromQL engine.

I also set the options argument to be inline code for the Markdown rendering.

The suggestion also avoids using the product name with a possessive which is recently identified area of our documentation style guide: https://developers.google.com/style/possessives#product,-feature,-and-company-names. I'm not certain we need to avoid it here but I do think Prometheus is a trademark so I thought better safe than sorry :) 

```suggestion
          "desc": "If set to `true` and the streaming engine is in use, fall back to using the Prometheus PromQL engine for a queries not supported by the streaming engine.",
```

charleskorn replied:
Thanks for the suggestion. I've changed this in d856405. 

I didn't use the inline code formatting for "true" as the option description is never rendered as Markdown anywhere.

pracucci commented on pkg/streamingpromql/errors.go:
```diff
@@ -7,8 +7,18 @@ import (
+	reason string
+}
+
+func NewNotSupportedError(reason string) error {
```
The reason passed here is quite high cardinality (because generated dinamically), and it's used as a label value. Have you considered introducing few classes of pre-defined reasons, and then keeping the current dynamic string as "details"? The reason will be used in metrics, the reason + details in the info log.

Examples of pre-defined reasons:
- unsupported aggregation (any aggregation)
- unsupported function (any function)
- unsupported function argument (any function, any argument)
- unsupported expression (any expression)
- offset
- grouping using without

charleskorn replied:
The maximum cardinality of this should be 109 series:

- There are 16 binary operators that we don't support, although at present these will be reported as one "binary operators not supported" series
- There are 11 aggregation functions that we don't support
- There are 64 functions that we don't support
- There are 10 aggregation over time functions that we don't support
- There are 8 particular cases we don't support that will have their own series:
  - instant vector selector with `offset`
  - range vector selector with `offset`
  - grouping with `without`
  - number literals as the top-level expression
  - string literals as the top-level expression
  - unary expressions
  - range vector selectors as the top-level expression
  - subqueries

I'm open to other opinions, but I'm inclined to keep this as-is to keep things simple and make analysis easier. As we implement support for each of these, the associated series will disappear, and we can use Adaptive Metrics to aggregate away unnecessary labels like `pod` etc. in the meantime. What do you think?

charleskorn replied:
I'm going to merge this as-is, but if you have any opinions on this @pracucci let me know and I can change the behaviour in a follow-up PR.

pracucci reviewed (approved):
LGTM (modulo a minor comment=

charleskorn commented:
Note that I've renamed the metrics since raising this PR: originally I used `cortex_querier_streaming_promql_engine_supported_queries_total`, but I've dropped the `querier` prefix in 291ea4a as these metrics could also potentially be emitted by query-frontends.

jhesketh reviewed (approved):
lgtm :-)
